### PR TITLE
Before DDP initialization, no need to place model to other device, wh…

### PR DESCRIPTION
**Pull request by sywangyi**
…ich will cause oom.

# What does this PR do?
llama 7b peft DDP out of memory after transformers upgrade to 4.31
Traceback (most recent call last):
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 551, in <module>
    main()
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 543, in main
    trainer.train(resume_from_checkpoint=training_args.resume_from_checkpoint)
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 507, in train
    return inner_training_loop(
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 860, in _inner_training_loop
    self.htcore.mark_step()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/utils/internal.py", line 26, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/step_closure.py", line 60, in mark_step
    htcore._mark_step(device_str)
RuntimeError: [Rank:5] FATAL ERROR :: MODULE:PT_SYNHELPER workspace Allocation of size ::9618705792 failed!
Traceback (most recent call last):
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 551, in <module>
    main()
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 543, in main
    trainer.train(resume_from_checkpoint=training_args.resume_from_checkpoint)
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 507, in train
    return inner_training_loop(
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 860, in _inner_training_loop
    self.htcore.mark_step()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/utils/internal.py", line 26, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/step_closure.py", line 60, in mark_step
    htcore._mark_step(device_str)
RuntimeError: [Rank:3] FATAL ERROR :: MODULE:PT_SYNHELPER workspace Allocation of size ::9618705792 failed!
Traceback (most recent call last):
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 551, in <module>
    main()
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 543, in main
    trainer.train(resume_from_checkpoint=training_args.resume_from_checkpoint)
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 507, in train
    return inner_training_loop(
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 860, in _inner_training_loop
    self.htcore.mark_step()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/utils/internal.py", line 26, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/step_closure.py", line 60, in mark_step
    htcore._mark_step(device_str)
RuntimeError: [Rank:1] FATAL ERROR :: MODULE:PT_SYNHELPER workspace Allocation of size ::9618705792 failed!
Traceback (most recent call last):
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 551, in <module>
    main()
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 543, in main
    trainer.train(resume_from_checkpoint=training_args.resume_from_checkpoint)
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 507, in train
    return inner_training_loop(
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 860, in _inner_training_loop
    self.htcore.mark_step()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/utils/internal.py", line 26, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/step_closure.py", line 60, in mark_step
    htcore._mark_step(device_str)
RuntimeError: [Rank:6] FATAL ERROR :: MODULE:PT_SYNHELPER workspace Allocation of size ::9618705792 failed!
Traceback (most recent call last):
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 551, in <module>
    main()
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 543, in main
    trainer.train(resume_from_checkpoint=training_args.resume_from_checkpoint)
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 507, in train
    return inner_training_loop(
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 860, in _inner_training_loop
    self.htcore.mark_step()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/utils/internal.py", line 26, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/step_closure.py", line 60, in mark_step
    htcore._mark_step(device_str)
RuntimeError: [Rank:2] FATAL ERROR :: MODULE:PT_SYNHELPER workspace Allocation of size ::9618705792 failed!
Traceback (most recent call last):
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 551, in <module>
    main()
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 543, in main
    trainer.train(resume_from_checkpoint=training_args.resume_from_checkpoint)
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 507, in train
    return inner_training_loop(
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 860, in _inner_training_loop
    self.htcore.mark_step()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/utils/internal.py", line 26, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/step_closure.py", line 60, in mark_step
    htcore._mark_step(device_str)
RuntimeError: [Rank:4] FATAL ERROR :: MODULE:PT_SYNHELPER workspace Allocation of size ::9618705792 failed!
Traceback (most recent call last):
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 551, in <module>
    main()
  File "/work/optimum-habana/examples/language-modeling/run_lora_clm.py", line 543, in main
    trainer.train(resume_from_checkpoint=training_args.resume_from_checkpoint)
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 507, in train
    return inner_training_loop(
  File "/work/optimum-habana/optimum/habana/transformers/trainer.py", line 860, in _inner_training_loop
    self.htcore.mark_step()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/utils/internal.py", line 26, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/step_closure.py", line 60, in mark_step
    htcore._mark_step(device_str)
RuntimeError: [Rank:7] FATAL ERROR :: MODULE:PT_SYNHELPER workspace Allocation of size ::9618705792 failed!

Fixes # (issue)
llama 7b peft DDP oom issue after transformers upgrade to 4.31

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
I just use following command to reproduce the issue in AWS DL1 instance. Gaudi2 may not reproduce this issue, since its memory is 96G
python ../gaudi_spawn.py \
    --world_size 8 --use_mpi run_lora_clm.py \
    --model_name_or_path huggyllama/llama-7b \
    --dataset_name tatsu-lab/alpaca \
    --bf16 True \
    --output_dir ./model_lora_llama \
    --num_train_epochs 3 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --gradient_accumulation_steps 4 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 2000 \
    --save_total_limit 1 \
    --learning_rate 1e-4 \
    --logging_steps 1 \
    --dataset_concatenation \
    --do_train \
    --use_habana \
    --use_lazy_mode \
    --throughput_warmup_steps 3

only out of memory in 8 cards using DDP. ok in single card.
if using device placement before DDP initialization:
============================================================================================
before DDP create = {'memory_allocated (GB)': 12.59, 'max_memory_allocated (GB)': 12.59, 'total_memory_available (GB)': 29.85}, process_idx = 2
before DDP create = {'memory_allocated (GB)': 12.59, 'max_memory_allocated (GB)': 12.59, 'total_memory_available (GB)': 29.85}, process_idx = 5
before DDP create = {'memory_allocated (GB)': 12.59, 'max_memory_allocated (GB)': 12.59, 'total_memory_available (GB)': 29.85}, process_idx = 7
before DDP create = {'memory_allocated (GB)': 12.59, 'max_memory_allocated (GB)': 12.59, 'total_memory_available (GB)': 29.85}, process_idx = 3
before DDP create = {'memory_allocated (GB)': 12.59, 'max_memory_allocated (GB)': 12.59, 'total_memory_available (GB)': 29.85}, process_idx = 1
before DDP create = {'memory_allocated (GB)': 12.59, 'max_memory_allocated (GB)': 12.59, 'total_memory_available (GB)': 29.85}, process_idx = 0
before DDP create = {'memory_allocated (GB)': 12.59, 'max_memory_allocated (GB)': 12.59, 'total_memory_available (GB)': 29.85}, process_idx = 6
before DDP create = {'memory_allocated (GB)': 12.59, 'max_memory_allocated (GB)': 12.59, 'total_memory_available (GB)': 29.85}, process_idx = 4
after ddp create = {'memory_allocated (GB)': 25.6, 'max_memory_allocated (GB)': 26.15, 'total_memory_available (GB)': 29.85},process_idx = 1
after ddp create = {'memory_allocated (GB)': 25.6, 'max_memory_allocated (GB)': 26.15, 'total_memory_available (GB)': 29.85},process_idx = 3
after ddp create = {'memory_allocated (GB)': 25.6, 'max_memory_allocated (GB)': 26.15, 'total_memory_available (GB)': 29.85},process_idx = 2
after ddp create = {'memory_allocated (GB)': 25.6, 'max_memory_allocated (GB)': 26.15, 'total_memory_available (GB)': 29.85},process_idx = 5
after ddp create = {'memory_allocated (GB)': 25.6, 'max_memory_allocated (GB)': 26.15, 'total_memory_available (GB)': 29.85},process_idx = 6
after ddp create = {'memory_allocated (GB)': 25.6, 'max_memory_allocated (GB)': 26.15, 'total_memory_available (GB)': 29.85},process_idx = 7
after ddp create = {'memory_allocated (GB)': 25.6, 'max_memory_allocated (GB)': 26.15, 'total_memory_available (GB)': 29.85},process_idx = 4
after ddp create = {'memory_allocated (GB)': 13.04, 'max_memory_allocated (GB)': 13.59, 'total_memory_available (GB)': 29.85},process_idx = 0

**Review by regisss — COMMENTED** [review]
Good catch! I just left one comment to clarify the case when doing evaluation only.

File: optimum/habana/accelerate/accelerator.py [hunk 1]
```diff
@@ -270,6 +270,8 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
         """
         if device_placement is None:
             device_placement = self.device_placement and self.distributed_type != DistributedType.FSDP
+            if not evaluation_mode and self.distributed_type == GaudiDistributedType.MULTI_HPU:
```
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/habana/accelerate/accelerator.py:273]: Makes sense :+1:

**Review by regisss — APPROVED** [review]
LGTM

**Comment by skaulintel** [comment]
I am enabling multi-card training for SDXL and this logic seems to be causing an issue for me. "RuntimeError: Got a non-HPU tensor, expecting an HPU tensor"

However, if I remove the model.to(device) call from the elif I am able to train successfully with gaudi_spawn.

**Comment by sywangyi** [comment]
Hi, device placement is in trainer.py. see https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L474

**Comment by skaulintel** [comment]
@sywangyi @regisss Yes - if I move the unet to device in our training script it seems to work. Thanks!

**Pull request merged**